File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/query/FilterField.java
package in.org.quicko.silo.client.query;

import java.util.List;

/**
 * One path awaiting an operator.
 *
 * <p>Equality is {@code isEqualTo} rather than the TypeScript client's
 * {@code equals}: a one-argument {@code equals} here would override
 * {@link Object#equals(Object)} instead of building a filter.
 */
public final class FilterField {
  private final String path;

  public FilterField(String path) {
    this.path = path;
  }

  public FilterExpression isEqualTo(Object value) {
    return leaf(FilterOperator.EQUALS, value);
  }

  public FilterExpression isNotEqualTo(Object value) {
    return leaf(FilterOperator.NOT_EQUALS, value);
  }

  public FilterExpression contains(Object value) {
    return leaf(FilterOperator.CONTAINS, value);
  }

  public FilterExpression greaterThan(Object value) {
    return leaf(FilterOperator.GREATER_THAN, value);
  }

  public FilterExpression atLeast(Object value) {
    return leaf(FilterOperator.AT_LEAST, value);
  }

  public FilterExpression lessThan(Object value) {
    return leaf(FilterOperator.LESS_THAN, value);
  }

  public FilterExpression atMost(Object value) {
    return leaf(FilterOperator.AT_MOST, value);
  }

  public FilterExpression oneOf(List<?> values) {
    return leaf(FilterOperator.IN, values);
  }

  public FilterExpression oneOf(Object... values) {
    return oneOf(List.of(values));
  }

  public FilterExpression exists() {
    return new FilterExpression(FilterNode.predicate(FilterOperator.EXISTS, path));
  }

  private FilterExpression leaf(FilterOperator operator, Object value) {
    return new FilterExpression(FilterNode.leaf(operator, path, value));
  }
}

File: packages/silo-client-dart/lib/src/query/filter_field.dart
import 'filter_expression.dart';
import 'filter_node.dart';
import 'filter_operator.dart';

/// One path awaiting an operator.
final class FilterField {
  const FilterField(this._path);

  final String _path;

  FilterExpression equals(Object? value) => _leaf(FilterOperator.equals, value);

  FilterExpression notEquals(Object? value) => _leaf(FilterOperator.notEquals, value);

  FilterExpression contains(Object? value) => _leaf(FilterOperator.contains, value);

  FilterExpression greaterThan(Object? value) => _leaf(FilterOperator.greaterThan, value);

  FilterExpression atLeast(Object? value) => _leaf(FilterOperator.atLeast, value);

  FilterExpression lessThan(Object? value) => _leaf(FilterOperator.lessThan, value);

  FilterExpression atMost(Object? value) => _leaf(FilterOperator.atMost, value);

  FilterExpression oneOf(List<Object?> values) => _leaf(FilterOperator.oneOf, values);

  FilterExpression exists() => FilterExpression(FilterNode(FilterOperator.exists, path: _path));

  FilterExpression _leaf(FilterOperator op, Object? value) =>
      FilterExpression(FilterNode(op, path: _path, value: value));
}
